systems-management/nisysmgmt.yml

```diff
@@ -0,0 +1,2076 @@
+        description: Internal errors that occured in the service.
+        items:
+          $ref: '#/definitions/Error'
+
+  Operation:
+    description: An operation provided by the API
+    type: object
+    properties:
+      available:
+        type: boolean
+        description: Whether the operation is available to the caller.
+        example: true
+      version:
+        type: integer
+        description: The version of the available operation.
+        example: 1
+    required: [available]
+
+  V1Operations:
+    description: V1 operations
+    type: object
+    properties:
+      operations:
+        description: >-
+          Available operations in the v1 version of the API:
+
+          - getJobs: The ability to retrieve the jobs
+          
+          - getSystems: The ability to retrieve the systems information
+
+          - getGroups: The ability to retrieve the groups
+    
+          - manageSystems: The ability to register/unregister systems
+          
+          - createOrCancelJobs: The ability to create or cancel jobs
+          
+          - activateSystems: The ability to activate systems
+          
+          - updateSystemMetadata: The ability to update systems metadata
+          
+          - manageFeeds: The ability to configure feeds and retrieve the available packages
+          
+          - createOrUpdateGroups: The ability to create, update or delete groups
+          
+        type: object
+        properties:
+          getJobs:
+            $ref: '#/definitions/Operation'
+          getSystems:
+            $ref: '#/definitions/Operation'
+          getGroups:
+            $ref: '#/definitions/Operation'
+          manageSystems:
+            $ref: '#/definitions/Operation'
+          createOrCancelJobs:
+            $ref: '#/definitions/Operation'
+          activateSystems:
+            $ref: '#/definitions/Operation'
+          updateSystemMetadata:
+            $ref: '#/definitions/Operation'
+          manageFeeds:
+            $ref: '#/definitions/Operation'
+          createOrUpdateGroups:
+            $ref: '#/definitions/Operation' 
+
+  QueryRequest:
+    description: Request for querying resources. More info here -> https://github.com/StefH/System.Linq.Dynamic.Core/wiki/Dynamic-Expressions and https://stefh.github.io/System.Linq.Dynamic.Core/html/144078bd-d1ad-bece-4daa-7fdd3a7ed7dc.htm
+    title: Query Request
+    properties:
+      skip:
+        description: How many resources to skip.
+        type: integer
+        default: 0
+        example: 100
+      take:
+        description: How many resources to take. Maximum value is 1000.
+        type: integer
+        default: -1
+        maximum: 1000
+        example: 100
+      filter:
+        type: string
+        example: connected.data.state==Connected
+        description: Specifies the filter criteria for resources.
+      projection:
+        type: string
+        example: 'new(id,grains,last_updated_timestamp)'
```

Comment:
Curious, why isn't this an array of strings or enums? Also, in your example, what does the `new` at the front mean (with parenthesis)?

Reply:
For now we don't want to restrict the API to certain keys when using filtering and projection(and I don't think we can do that since we have the grains dictionary that can contain anything). So we let the user to send a string that should be built based on the JobModel/SystemModel properties. We will make sure that the models aren't changed so we will avoid breaking the API. The "new" operator implies creation of a new object(dict). Without that, the query will return a tuple with the values instead of a dict. So the "new" operator is very important when you want to keep the hierarchy of the response and to avoid flattening. This way the client doesn't have to write custom logic to handle the query response for each different query request.